### PR TITLE
Add containers by hostname view.

### LINKED ISSUE #544: Add containers-by-hostname view to scope
This is better than container-by-image, as it approximates services more closely (containers with the same host name being considered replicas for the same service)

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #544

### CODE REVIEW

**Comment by tomwilkie** [comment]
@davkal what do you think re:UX?  Maybe make the sub topologies a drop down if there are more than 1?

screen shot 2015-10-05 at 16 04 42

**Comment by davkal** [comment]
-1 on dropdowns, especially if it's only for 2 items.

Long term we could have groupings in one line like the time filter in google finance charts:
screen shot 2015-10-09 at 15 21 32

**Comment by tomwilkie** [comment]
Fair enough; are you happy with UX as is?

**Comment by davkal** [comment]
Yup. Sorry, should have said that.

**Comment by tomwilkie** [comment]
How about, as you hover over Containers, the different groupings appear?  Then if you select one it stays visible?

**Comment by davkal** [comment]
Hover only works with mouse. Leaves out ipad et al. for cool demos.

**Comment by davkal** [comment]
Or not on hover, but click on main topology will show sub topologies. That would work on tablets too.

**Comment by tomwilkie** [comment]
Ah yes.  Good point.  How about, on tablets, we use a 3d UI?

File: probe/docker/container.go [hunk 1]
```diff
@@ -221,6 +222,7 @@ func (c *container) GetNode(localAddrs []net.IP) report.Node {
 		ImageID:          c.container.Image,
 		ContainerIPs: strings.Join(append(c.container.NetworkSettings.SecondaryIPAddresses,
 			c.container.NetworkSettings.IPAddress), " "),
+		ContainerHostname: c.container.Config.Hostname,
```
Comment (paulbellamy) [inline probe/docker/container.go]: This should include the domainname.

File: render/mapping.go [hunk 2]
```diff
@@ -594,6 +594,28 @@ func MapContainer2Pod(n RenderableNode, _ report.Networks) RenderableNodes {
...
 
+// MapContainer2Hostname maps container RenderableNodes to 'hostname' renderabled nodes..
+func MapContainer2Hostname(n RenderableNode, _ report.Networks) RenderableNodes {
+	// Propogate all pseudo nodes
+	if n.Pseudo {
+		return RenderableNodes{n.ID: n}
+	}
+
+	// Otherwise, if some some reason the container doesn't have a hostname
+	// (maybe slightly out of sync reports), just drop it
```
Comment (paulbellamy) [inline render/mapping.go]: Is that really a reason the container wouldn't have a hostname?

File: render/mapping.go [hunk 3]
```diff
@@ -594,6 +594,28 @@ func MapContainer2Pod(n RenderableNode, _ report.Networks) RenderableNodes {
...
+	id, ok := n.Node.Metadata[docker.ContainerHostname]
+	if !ok {
+		return RenderableNodes{}
+	}
+
+	// Add container-<id> key to NMD, which will later be counted to produce the minor label
```
Comment (paulbellamy) [inline render/mapping.go]: This comment is not what we are doing.

File: render/topologies.go [hunk 4]
```diff
@@ -171,6 +171,25 @@ var ContainerImageRenderer = Map{
...
+				MapFunc:  MapContainer2ContainerImage,
+				Renderer: ContainerRenderer,
+			},
+			Map{
+				MapFunc:  MapContainerImageIdentity,
+				Renderer: SelectContainerImage,
+			},
+		),
+	},
+}
```
Comment (paulbellamy) [inline render/topologies.go]: Are we just grouping containers by hostname, or by container image, then hostname? It looks like the latter, but checking whether that is what we want.

File: render/mapping.go (hunk 2 above)
Comment (tomwilkie) [inline render/mapping.go]: Yes; if the containers collect gets out of sync for some reason with the endpoint collection, you can end up with a container node with little metadata.  We can safely ignore them in this case.

File: render/topologies.go (hunk 4 above)
Comment (tomwilkie) [inline render/topologies.go]: Yeah this is a bad merge I think.

**Comment by tomwilkie** [comment]
@paulbellamy PTAL

**Comment by paulbellamy** [comment]
LGTM

**Pull request merged**